src/jingle_sip/mod_jingle_sip_backend.erl

```diff
@@ -0,0 +1,199 @@
+-module(mod_jingle_sip_backend).
+
+-include("mongoose.hrl").
+
+-type call_id() :: binary().
+-type incoming_request() :: {node(), binary()}.
+-type outgoing_handle() :: binary().
+
+-export([init/2]).
+-export([set_incoming_request/5]).
+-export([set_incoming_handle/2]).
+-export([set_outgoing_request/4]).
+-export([set_outgoing_handle/4]).
+-export([set_outgoing_accepted/1]).
+-export([set_incoming_accepted/1]).
+-export([get_incoming_request/1]).
+-export([get_outgoing_handle/1]).
+-export([get_session_info/1]).
+-export([remove_session/1]).
+
+-record(jingle_sip_session, {sid,
+                             dialog,
+                             state,
+                             direction,
+                             request,
+                             node,
+                             meta}).
+
+init(_Host, _Opts) ->
+    mnesia:create_table(jingle_sip_session,
+                        [{ram_copies, [node()]},
+                         {attributes, record_info(fields, jingle_sip_session)}]).
+
+set_incoming_request(CallID, ReqID, From, To, JingleEl) ->
+    TFun = pa:bind(fun set_incoming_request_tr/5, CallID, ReqID, From, To, JingleEl),
+    run_transaction(TFun).
+
+set_incoming_request_tr(CallID, ReqID, _From, _To, JingleEl) ->
+    case mnesia:wread({jingle_sip_session, CallID}) of
+        [#jingle_sip_session{request = undefined, direction = in, meta = Meta} = Session] ->
+            NewMeta = Meta#{init_stanza => JingleEl},
+            Session2 = Session#jingle_sip_session{request = ReqID,
+                                                  node = node(),
+                                                  meta = NewMeta
+                                                  },
+            mnesia:write(Session2);
+        [_] ->
+            {error, sid_already_exists};
+        _ ->
+            Meta = #{init_stanza => JingleEl},
+            Session = #jingle_sip_session{sid = CallID,
+                                          request = ReqID,
+                                          dialog = undefined,
+                                          state = undefined,
+                                          direction = in,
+                                          node = node(),
+                                          meta = Meta},
+            mnesia:write(Session)
+    end.
+
+set_outgoing_request(CallID, ReqID, From, To) ->
+    TFun = pa:bind(fun set_outgoing_request_tr/4, CallID, ReqID, From, To),
```

Comment:
The version with `fun() -> ... end` would still be dialyzer type transparent. This means you can find more potential errors with dialyzer.

```erlang
TFun = fun() -> set_outgoing_request_tr(CallID, ReqID, From, To) end,
run_transaction(TFun).
```

Using `pa` can be useful, if we want to use partial application, which is not the case here.

Reply:
I simply try to avoid unnamed funs, that's the reason I used `pa` here.